code/player/controllers/body_mechanics/AntiStuck.cs

```diff
@@ -0,0 +1,81 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+
+using Sandbox;
+
+namespace Cinema;
+
+public partial class AntiStuckMechanic : PlayerBodyControllerMechanic
+{
+    internal int UnstuckAttempts = 0;
+
+    protected override void Simulate()
+    {
+        if( !IsStuck() )
+        {
+            //Reset unstuck attempts if they aren't actually stuck
+            UnstuckAttempts = 0;
+            safePos = Vector3.Zero;
+        }
+        else
+        {
+            AttemptUnstuck();
+        }
+    }
+
+    //Check if the player is stuck
+    protected bool IsStuck()
+    {
+        var traceBox = Controller.TraceBBox(Controller.Position, Controller.Position);
+
+        return traceBox.StartedSolid;
+    }
+
+    protected bool AttemptUnstuck()
+    {
+        //Only the server should unstuck the player
+        if (Game.IsClient) return true;
+
+        int triesPerTick = 20;
+
+        for (int i = 0; i < triesPerTick; i++)
+        {
+            //var pos = Controller.Position + Vector3.Random.Normal * (UnstuckAttempts / 2.0f);
+            var pos = Controller.Position + (Vector3.Random * 45 * UnstuckAttempts);
+            DebugOverlay.Line(Controller.Position, pos, 5.0f);
+
+            var check = Controller.TraceBBox(pos, Controller.Position);
+            if (ValidatePosition(check))
+            {
+                Controller.Position = safePos;
+                return false;
+            }
+
+            check = Controller.TraceBBox(pos, pos + Vector3.Down * 64.0f);
+            if (ValidatePosition(check))
+            {
+                Controller.Position = safePos;
+                return false;
+            }
+        }
+
+        UnstuckAttempts++;
+
+        return true;
+    }
+
+    Vector3 safePos;
+
+    protected bool ValidatePosition(TraceResult result)
+    {
+        if (result.StartedSolid) return false;
+        if (result.Normal.z != 1) return false;
+
+        safePos = result.HitPosition;
```

Comment:
This is an anti-pattern which causes the code to become confusing to read:

If you look at what calls this `AttemptUnstuck` we see that a trace `check` is validated with this function. If the trace is "valid" ie. did not start solid and the normal's z-axis is not 1, then the trace is valid and a boolean true is returned.

However this function causes a mutation of state when true is returned, `safePos` is set to the hitposition of the validated trace.

From an outside perspective of just looking at the `AttemptUnstuck` method this mutation of state is very confusing since it reads as such:

```cs
var check = someTrace;
if (ValidatePosition(check)) {
  Controller.Position = safePos;
}
```

The reader will immediately wonder, where did `safePos` come from and how does it relate to `check`'s validity?

The proper pattern here is to just return true, do not mutate the state (thus this is a pure/static function) and in the `AttemptUnstuck` function the code should read more like: 

```cs
var check = someTrace;
if (ValidatePosition(check)) {
  Controller.Position = check.HitPosition;
}
```

Furthermore there is no need for `safePos` to exist.